docopt/__init__.py

```diff
@@ -709,39 +709,101 @@ def isanumber(x):
     return parsed
 
 
-def parse_defaults(docstring: str) -> list[Option]:
-    defaults = []
-    for s in parse_section("options:", docstring):
-        options_literal, _, s = s.partition(":")
-        if " " in options_literal:
-            _, _, options_literal = options_literal.partition(" ")
-        assert options_literal.lower().strip() == "options"
-        split = re.split(r"\n[ \t]*(-\S+?)", "\n" + s)[1:]
-        split = [s1 + s2 for s1, s2 in zip(split[::2], split[1::2])]
-        for s in split:
-            if s.startswith("-"):
-                arg, _, description = s.partition("  ")
-                flag, _, var = arg.replace("=", " ").partition(" ")
-                option = Option.parse(s)
-                defaults.append(option)
-    return defaults
-
-
-def parse_section(name: str, source: str) -> list[str]:
-    pattern = re.compile(
-        "^([^\n]*" + name + "[^\n]*\n?(?:[ \t].*?(?:\n|$))*)",
-        re.IGNORECASE | re.MULTILINE,
+class DocSections(NamedTuple):
+    before_usage: str
+    usage_header: str
+    usage_body: str
+    after_usage: str
+
+
+def parse_docstring_sections(docstring: str) -> DocSections:
+    """Partition the docstring into the main sections.
+
+    The docstring is returned, split into a tuple of 4 pieces: text before the
+    usage section, the usage section header, the usage section body and text
+    following the usage section.
+    """
+    usage_pattern = r"""
+    # Any number of lines precede the usage section
+    \A(?P<before_usage>(?:.*\n)*?)
```

Comment:
Isn't the `?` redundant in the `*?` ? can't we just have `*`?

Reply:
Ah, so this is a bit tricky. I thought the same looking at the original regexes docopt used, I assumed `+?` could just be `*`. But the `?` acts as a modifier to make the `*` non-greedy . So the effect here is that we'll match as few lines as possible before a `usage:` line occurs. 

Without the non-greedy operator it'd match as many lines as possible. This is actually still OK for correctness, as the match will initially consume all the lines, then backtrack until it finds a `usage:` line to match the usage section with, and succeed from there. However the effect of that is that the final usage: line in the doc will be used. With the non-greedy match, the first usage: section gets used.

So it only matters if a doc incorrectly has multiple usage sections. There is some code that tries to check for multiple usage sections, and raises an error if there are multiple (that's `lint_docstring` in this version). I just tried with the standard greedy match, and `lint_docstring` currently fails to notice the second usage section in this case, as it currently looks only in the usage body and the section after usage, not in the section before.

The same effect as the non-greedy match can be achieved in this case using a negative lookahead assertion in this `before_usage` group to not allow it to match lines containing `usage:`. That seems like a more clear approach, as non-greedy matching is pretty obscure.

Reply:
Thank you, that makes sense. I think you're right, the negative lookahead is more clear, otherwise I'd want a comment here, and it would be great to not have a comment.

Reply:
- [ ] make usage_pattern regex more intuitive by using lookahead instead of non-greedy match